gwin/models/gaussian_noise.py

```diff
@@ -499,59 +501,213 @@ class MarginalizedPhaseGaussianNoise(GaussianNoise):
-                # calculate inner products
                 hh_i = h[self._kmin:kmax].inner(h[self._kmin:kmax]).real
-                hd_i = self.data[det][self._kmin:kmax].inner(
-                    h[self._kmin:kmax])
-            # store
+                if self._margtime:
+                    hd_i = 4. * (h.delta_f) * numpy.fft.fft(
+                               numpy.conj(self.data[det][self._kmin:kmax]) *
+                               h[self._kmin:kmax]).real
+                else:
+                    hd_i = self.data[det][self._kmin:kmax].inner(
+                           h[self._kmin:kmax])
+            opt_snr += hh_i
+            mf_snr += hd_i
             setattr(self._current_stats, '{}_optimal_snrsq'.format(det), hh_i)
-            # TODO: is it possible to store a something like cplx_loglr for
-            # each detector?
-            hh += hh_i
-            hd += hd_i
-        hd = abs(hd)
-        return numpy.log(special.i0e(hd)) + hd - 0.5*hh
+            if self._margdist or self._margphi:
+                setattr(self._current_stats,
+                        '{}_matchedfilter_snrsq'.format(det), hd_i)
+        mf_snr = abs(mf_snr)
+        if self._margtime and self._margdist and self._margphi:
+            # log likelihood marginalised over time and phase. The phase
+            # marginalization adds the Bessel function and the time
+            # marginalization adds the Fourier transform
+            logl = special.logsumexp(numpy.log(special.i0(mf_snr)),
+                                     b=delta_t)
+            logl_marg = logl/dist_array
+            opt_snr_marg = opt_snr/dist_array**2
+            return special.logsumexp(logl_marg - 0.5*opt_snr_marg, b=delta_d)
+        elif self._margtime and self._margdist:
+            # log likelihood marginalized over time
+            logl = special.logsumexp(mf_snr, b=delta_t)
+            logl_marg = logl/dist_array
+            opt_snr_marg = opt_snr/dist_array**2
+            return special.logsumexp(logl_marg - 0.5*opt_snr_marg,
+                                     b=delta_d)
+        elif self._margtime and self._margphi:
+            return special.logsumexp(numpy.log(special.i0(mf_snr)),
+                                     b=delta_t) - 0.5*opt_snr
+        elif self._margdist and self._margphi:
+            # log likelihood marginalizaled over phase
+            logl = numpy.log(special.i0(mf_snr))
+            logl_marg = logl/dist_array
+            opt_snr_marg = opt_snr/dist_array**2
+            return special.logsumexp(logl_marg - 0.5*opt_snr_marg, b=delta_d)
+        elif self._margdist:
+            mf_snr_marg = mf_snr/dist_array
+            opt_snr_marg = opt_snr/dist_array**2
+            return special.logsumexp(mf_snr_marg - 0.5*opt_snr_marg, b=delta_d)
+        elif self._margtime:
+            return special.logsumexp(mf_snr, b=delta_t) - 0.5*opt_snr
+        return numpy.log(special.i0(mf_snr)) - 0.5*opt_snr
```

Comment:
Define a different hidden function for each of these possibilities (though you may be able to nest some of these), then move all the if/else's into the `__init__`, setting some attribute to the correct function. For example, in the `__init__`:
```
def __init__(...):
    <stuff>
    if self._margtime and self._margdist and self._margphi:
        self._evalmfsnr = self._margtimephasedist
    elif self._evalmargtime and self._margdist:
        self._evalmfsnr = self._margtimedist
    <etc>
```
Then in the loglr function at line 682 you'd just have:
```
return self._evalmfsnr(mf_snr)
```
That'd make the loglr function a bit easier to read, and make it so that all of the logic evaluation only happens once.